### ISSUE TITLE
Hangfire throws ObjectDisposedException during twice Dispose

### ISSUE THREAD

**Issue by dawidlazuk** [HangfireIO/Hangfire#2097]
In our integration tests, we use MassTransit and Hangfire. Unfortunately disposing of MassTransit includes disposing of all the hosted services found in the process, including the Hangfire services. Afterward Hangfire tries to dispose itself again resulting in `System.ObjectDisposedException : Cannot access a disposed object. Object name: 'Hangfire.Server.BackgroundProcessingServer'.`

It is obvious that the responsibility of the issue is on the MassTransit disposing of things out of its scope but IMO it would be better to make Hangfire Dispose operation idempotent instead of throwing exceptions.

**Comment by odinserj** [comment]
Could you please post here full stack trace of the `ObjectDisposedException`?

**Comment by dawidlazuk** [comment]
```
  Error Message:
   TearDown : System.ObjectDisposedException : Cannot access a disposed object.
Object name: 'Hangfire.Server.BackgroundProcessingServer'.
  Stack Trace:
  --TearDown
   at Hangfire.Server.BackgroundProcessingServer.ThrowIfDisposed()
   at Hangfire.Server.BackgroundProcessingServer.SendStop()
   at Hangfire.BackgroundJobServer.SendStop()
   at Hangfire.BackgroundJobServerHostedService.StopAsync(CancellationToken cancellationToken)
   at MassTransit.DependencyInjection.Testing.ContainerTestHarness.<DisposeAsync>b__15_0(IHostedService x) in /_/src/MassTransit/DependencyInjection/DependencyInjection/Testing/ContainerTestHarness.cs:line 58
   at System.Linq.Enumerable.SelectArrayIterator`2.MoveNext()
   at System.Threading.Tasks.Task.WhenAll(IEnumerable`1 tasks)
   at MassTransit.DependencyInjection.Testing.ContainerTestHarness.DisposeAsync() in /_/src/MassTransit/DependencyInjection/DependencyInjection/Testing/ContainerTestHarness.cs:line 86
   at Autofac.Core.Disposer.DisposeAsync(Boolean disposing)
   at Autofac.Core.Lifetime.LifetimeScope.DisposeAsync(Boolean disposing)
   at Autofac.Core.Container.DisposeAsync(Boolean disposing)
   at Autofac.Extensions.DependencyInjection.AutofacServiceProvider.DisposeAsync()
   at Microsoft.AspNetCore.Hosting.WebHost.DisposeServiceProviderAsync(IServiceProvider serviceProvider)
   at Microsoft.AspNetCore.Hosting.WebHost.DisposeAsync()
   at Microsoft.AspNetCore.Hosting.WebHost.Dispose()
   at OurService.IntegrationTests.Sagas.SomeStateMachineTests.TearDown() in /home/runner/work/OurService/OurService/OurService.IntegrationTests/Sagas/SomeStateMachineTests.cs:line 414
   at NUnit.Framework.Internal.TaskAwaitAdapter.GenericAdapter`1.GetResult()
   at NUnit.Framework.Internal.AsyncToSyncAdapter.Await(Func`1 invoke)
   at NUnit.Framework.Internal.Commands.SetUpTearDownItem.RunSetUpOrTearDownMethod(TestExecutionContext context, IMethodInfo method)
   at NUnit.Framework.Internal.Commands.SetUpTearDownItem.RunTearDown(TestExecutionContext context)
```

**Comment by odinserj** [comment]
`Dispose` methods in Hangfire can be called multiple times, but when another method is called after instance is disposed, like `SendStop`, it should throw the exception, and this is a standard practice. You can remove a call to `SendStop` from your `DisposeAsync` implementation and call only `Dispose` there – Hangfire checks if stop method is called when disposing a server instance and will trigger it automatically otherwise to avoid waiting infinitely.

**Comment by shaynevanasperen** [comment]
I'm also facing this problem. But I'm not even using MassTransit. I'm simply using the `WebApplicationFactory<T>` from Microsoft which automatically calls `DisposeAsync` when the test is finished running. This is the stack trace:

```
System.AggregateException: One or more hosted services failed to stop. (Cannot access a disposed object.
Object name: 'Hangfire.Server.BackgroundProcessingServer'.)
 ---> System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Hangfire.Server.BackgroundProcessingServer'.
   at Hangfire.Server.BackgroundProcessingServer.ThrowIfDisposed()
   at Hangfire.Server.BackgroundProcessingServer.SendStop()
   at Hangfire.BackgroundJobServer.SendStop()
   at Hangfire.BackgroundJobServerHostedService.StopAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StopAsync(CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.Hosting.Internal.Host.StopAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.DisposeAsync()
```

@odinserj Your suggestion to remove a call to `SendStop` is not possible, as the code is buried deep inside the `Microsoft.AspNetCore.Mvc.Testing` library.

This is definitely a bug in `Hangfire.AspNetCore`.

**Comment by kcadduk** [comment]
Did you happen to fix this issue? I am having the same problem. Masstransit and Hangfire seeming to conflict with each other.

**Comment by bfondu** [comment]
Having the same issue as @shaynevanasperen with WebApplicationFactory tests. 
The issue very much appears to be timing related, as it doesn't happen 100% of the time, and doesn't happen when debugging. Might be a concurrency issue around `_stoppingCts`?

For now, I've implemented xUnit's IAsyncLifetime in an inherited WebApplicationFactory, and put a try/catch around the dispose:
```public new async Task DisposeAsync()
    {
        try
        {
            await base.DisposeAsync();
        }
        catch (ObjectDisposedException)
        {
        }
    }```
    
A dirty hack for sure, but sadly seems to be necessary to have reliable tests.

**Comment by shaynevanasperen** [comment]
@odinserj I don't see how that change will prevent an `ObjectDisposedException` from beng thrown. Did you confirm that it actually works? Were you even able to reproduce it yourself? Shouldn't there be some sort of test coverage for this?

**Comment by odinserj** [comment]
It's making the `_server` field `null` after disposing the object. So when MassTransit's `ContainerTestHarness` class will try to call its [`StopAsync`](https://github.com/MassTransit/MassTransit/blob/7d14dbd9cb7de15561f95a5e73f587e6fd9d9068/src/MassTransit/DependencyInjection/DependencyInjection/Testing/ContainerTestHarness.cs#L63) method after `Dispose` method if called, `_server` field will be already `null` and nothing will be called on it.

**Comment by adar2** [comment]
@odinserj, I'm experiencing the same issue with ```BackgroundJobServerHostedService```.
Could you make the ```_processingServer``` field ```null``` after you call ```Dispose``` there as well?

**Comment by wrakocy** [comment]
> I'm also facing this problem. But I'm not even using MassTransit. I'm simply using the `WebApplicationFactory<T>` from Microsoft which automatically calls `DisposeAsync` when the test is finished running. This is the stack trace:
> 
> ```
> System.AggregateException: One or more hosted services failed to stop. (Cannot access a disposed object.
> Object name: 'Hangfire.Server.BackgroundProcessingServer'.)
>  ---> System.ObjectDisposedException: Cannot access a disposed object.
> Object name: 'Hangfire.Server.BackgroundProcessingServer'.
>    at Hangfire.Server.BackgroundProcessingServer.ThrowIfDisposed()
>    at Hangfire.Server.BackgroundProcessingServer.SendStop()
>    at Hangfire.BackgroundJobServer.SendStop()
>    at Hangfire.BackgroundJobServerHostedService.StopAsync(CancellationToken cancellationToken)
>    at Microsoft.Extensions.Hosting.Internal.Host.StopAsync(CancellationToken cancellationToken)
>    --- End of inner exception stack trace ---
>    at Microsoft.Extensions.Hosting.Internal.Host.StopAsync(CancellationToken cancellationToken)
>    at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.DisposeAsync()
> ```
> 
> @odinserj Your suggestion to remove a call to `SendStop` is not possible, as the code is buried deep inside the `Microsoft.AspNetCore.Mvc.Testing` library.
> 
> This is definitely a bug in `Hangfire.AspNetCore`. I'm using the latest stable release: `1.7.33`.

I've just updated to the latest (1.8.3) and am still having this issue. @bfondu's IAsyncLifetime implementation does not work for me. Note that, as @bfondu notes, the error occurs only intermittently (in my case, from within an ADO pipeline), and does not seem to locally. 

Any suggestions?

Here is my stack trace:

  Error Message:
   [Test Class Cleanup Failure (XXX.XXXX.FunctionalTests.Pages.ReverseWires.Dashboard)]: System.AggregateException : One or more hosted services failed to stop. (Cannot access a disposed object.
Object name: 'Hangfire.Server.BackgroundProcessingServer'.)
---- System.ObjectDisposedException : Cannot access a disposed object.
Object name: 'Hangfire.Server.BackgroundProcessingServer'.
  Stack Trace:
     at Microsoft.Extensions.Hosting.Internal.Host.StopAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.DisposeAsync()
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.Dispose(Boolean disposing)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.Dispose()
----- Inner Stack Trace -----
   at Hangfire.Server.BackgroundProcessingServer.ThrowIfDisposed()
   at Hangfire.Server.BackgroundProcessingServer.SendStop()
   at Hangfire.BackgroundJobServer.SendStop()
   at Hangfire.BackgroundJobServerHostedService.StopAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StopAsync(CancellationToken cancellationToken)

**Comment by wrakocy** [comment]
I wound up using a slight variation of @bfondu 's fix to work around this. 

Instead of implementing IAsyncLifetime in my concrete WebApplicationFactory, I implemented both IDisposable and IAsyncDisposable and trapped for AggregateExceptions with single inner exceptions of type ObjectDisposedException. 

Not pretty, but seems to be working:
  
```
  public new void Dispose()
  {
      try
      {
          base.Dispose();
      }
      catch (AggregateException ex) 
      {
          var exs = ex.InnerExceptions;

          if(exs?.Count == 1)
          {
              if (exs.Single() is ObjectDisposedException)
                  return;
          }

          throw;
      }
  }

  public new Task DisposeAsync()
  {
      Dispose();
      return Task.CompletedTask;
  }
```

**Comment by odinserj** [comment]
@wrakocy, you are right – this problem was fixed, but unfortunately only for `BackgroundProcessingServerHostedService` (new in 1.8.3), not for the `BackgroundJobServerHostedService` class. Today I'm releasing 1.7.36 with this problem fixed for 1.7.X, but 1.8.4 will be released later – perhaps the simplest way to use the `1.8.3-build-03811` version from the CI feed.

https://ci.appveyor.com/nuget/hangfire
